pkg/mimirpb/custom.go

```diff
@@ -6,19 +6,44 @@ import (
 	"bytes"
 	"fmt"
 	"math"
+	"syscall"
+	"testing"
+	"unsafe"
 
 	gogoproto "github.com/gogo/protobuf/proto"
 	"github.com/prometheus/prometheus/model/histogram"
+	"go.uber.org/atomic"
 	"google.golang.org/grpc/encoding"
 	"google.golang.org/grpc/encoding/proto"
 	"google.golang.org/grpc/mem"
 	protobufproto "google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/protoadapt"
 )
 
+type CustomCodecConfig struct {
+	InstrumentRefLeaksPct float64
+}
+
+func (cfg CustomCodecConfig) Codec() encoding.CodecV2 {
+	c := &codecV2{codec: encoding.GetCodecV2(proto.Name)}
+	if cfg.InstrumentRefLeaksPct > 0 {
+		c.instrumentRefLeaksOneIn = uint64(math.Trunc(100 / cfg.InstrumentRefLeaksPct))
+	}
+	return c
+}
+
+var GlobalCodec encoding.CodecV2
+
+func (cfg CustomCodecConfig) RegisterGlobally() {
+	GlobalCodec = cfg.Codec()
+	encoding.RegisterCodecV2(GlobalCodec)
+}
+
 func init() {
-	c := encoding.GetCodecV2(proto.Name)
-	encoding.RegisterCodecV2(&codecV2{codec: c})
+	// Instrument all buffers when testing.
+	if testing.Testing() {
+		CustomCodecConfig{InstrumentRefLeaksPct: 100}.RegisterGlobally()
+	}
```

Comment:
Please be very careful of edge cases when moving the codec registration from a package `init` function to `mimir.New`. Since `init` is transparently called on import, have we checked that _any entry point, anywhere in the system_ that imports this package still calls `RegisterGlobally()`?

If we're not 100% sure, it might gain a layer of safety to just register `codecV2` with 0 instrumented buffers here, so we're sure we continue using it in all contexts, and then update the value at the time config is set?